File: .github/cookbook/copilot-sdk/go/recipe/persisting-sessions.go
package main

import (
	"context"
	"fmt"
	"log"

	copilot "github.com/github/copilot-sdk/go"
)

func main() {
	ctx := context.Background()
	client := copilot.NewClient(nil)
	if err := client.Start(ctx); err != nil {
		log.Fatal(err)
	}
	defer client.Stop()

	// Create session with a memorable ID
	session, err := client.CreateSession(ctx, &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		SessionID:           "user-123-conversation",
		Model:               "gpt-5.4",
	})
	if err != nil {
		log.Fatal(err)
	}

	_, err = session.SendAndWait(ctx, copilot.MessageOptions{Prompt: "Let's discuss TypeScript generics"})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Session created: %s\n", session.SessionID)

	// Disconnect session but keep data on disk
	session.Disconnect()
	fmt.Println("Session disconnected (state persisted)")

	// Resume the previous session
	resumed, err := client.ResumeSession(ctx, "user-123-conversation", &copilot.ResumeSessionConfig{OnPermissionRequest: copilot.PermissionHandler.ApproveAll})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Resumed: %s\n", resumed.SessionID)

	_, err = resumed.SendAndWait(ctx, copilot.MessageOptions{Prompt: "What were we discussing?"})
	if err != nil {
		log.Fatal(err)
	}

	// List sessions
	sessions, err := client.ListSessions(ctx, nil)
	if err != nil {
		log.Fatal(err)
	}
	ids := make([]string, 0, len(sessions))
	for _, s := range sessions {
		ids = append(ids, s.SessionID)
	}
	fmt.Printf("Sessions: %v\n", ids)

	// Delete session permanently
	if err := client.DeleteSession(ctx, "user-123-conversation"); err != nil {
		log.Fatal(err)
	}
	fmt.Println("Session deleted")

	resumed.Disconnect()
}

File: .github/cookbook/copilot-sdk/python/recipe/persisting_sessions.py
#!/usr/bin/env python3

import asyncio
from copilot import CopilotClient, SessionConfig, MessageOptions, PermissionHandler

async def main():
    client = CopilotClient()
    await client.start()

    # Create session with a memorable ID
    session = await client.create_session(SessionConfig(
        session_id="user-123-conversation",
        model="gpt-5",
        on_permission_request=PermissionHandler.approve_all))

    await session.send_and_wait(MessageOptions(prompt="Let's discuss TypeScript generics"))
    print(f"Session created: {session.session_id}")

    # Destroy session but keep data on disk
    await session.destroy()
    print("Session destroyed (state persisted)")

    # Resume the previous session
    resumed = await client.resume_session("user-123-conversation", on_permission_request=PermissionHandler.approve_all)
    print(f"Resumed: {resumed.session_id}")

    await resumed.send_and_wait(MessageOptions(prompt="What were we discussing?"))

    # List sessions
    sessions = await client.list_sessions()
    print("Sessions:", [s.session_id for s in sessions])

    # Delete session permanently
    await client.delete_session("user-123-conversation")
    print("Session deleted")

    await resumed.destroy()
    await client.stop()

if __name__ == "__main__":
    asyncio.run(main())
